addons/udes_sale_stock/models/sale_order_line.py

```diff
@@ -22,6 +22,78 @@ class SaleOrderLine(models.Model):
         default=False,
         index=True,
     )
+    u_delivery_line_state = fields.Selection(
+        [
+            ("confirmed", "Confirmed"),
+            ("no_stock", "No Stock"),
+            ("allocated", "Allocated"),
+            ("picked", "Picked"),
+            ("packed", "Packed"),
+            ("done", "Done"),
+            ("cancelled", "Cancelled"),
+        ],
+        string = "Delivery Status",
+        compute = "_find_delivery_line_status",
+        store = True,
+
+    )
+
+    @api.depends("move_ids", "move_ids.state", "product_id", "order_id.state")
+    def _find_delivery_line_status(self):
+        """
+        Added as a part of SE-1974
+        - Sale order line will always have route_id field set.
+        - There are 2 new field added on stock.location.route
+        u_pick_operation_type_ids, u_pack_operation_type_ids these should be correctly set.
+        - We will first look at the pick pickings and set status according to it.
+        - Then we will look at the pack pickings and set status according to it.
+        - States:
+        "Confirmed: Pickings are created and in waiting state"
+        "No Stock: One or more line has no stock"
+        "Allocated: All pick pickings are reserved or in done state"
+        "Picked: All pick pickings are in done state"
+        "Packed: All pack pickings are in done state"
+        "Cancelled: Sale order is in cancel state"
+        "Done: All pickings are in done state"
+        """
+        for line in self:
+            state = False
+            if line.route_id:
+                route_id = line.route_id
+                pick_op_type = route_id.u_pick_operation_type_ids
+                pack_op_type = route_id.u_pack_operation_type_ids
+                if not pick_op_type and not pack_op_type:
+                    line.u_delivery_line_state = state
+                    continue
+                if line.order_id.state == "cancel":
+                    state = "cancelled"
+                    line.u_delivery_line_state = state
```

Comment:
nit: if we are doing `continue` we can probably skip assignment to `state` variable and just assign "cancelled" directly here, and similar change for other branches which end with `continue`.